cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -250,7 +260,53 @@ public void getAllApps(HttpRequest request, HttpResponder responder,
       .map(ApplicationRecord::new)
       .collect(Collectors.toList());
 
-    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(apps));
+    if (Optional.ofNullable(pageSize).orElse(0) != 0) {
+      PaginatedApplicationRecords result = getPaginatedResults(apps, pageToken, pageSize, orderBy,
+                                                               nameFilter);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(result));
+    } else {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(apps));
+    }
+  }
+
+  /**
+   * TODO: This is a temporary implementation, will be deprecated after
+   * https://cdap.atlassian.net/browse/CDAP-18603 is fully implemented.
+   * Returns paginated result for getApps, also includes nextPageToken
+   *
+   * @param apps the applications obtained from the Store
+   * @param pageToken the next pageToken to return the paginated results from
+   * @param pageSize the total number of application records to display on the result page
+   * @param orderBy  ASC for ascending order, DESC for descending order
+   * @param nameFilter filter applications whose name contains nameFilter
+   */
+  private PaginatedApplicationRecords getPaginatedResults(List<ApplicationRecord> apps,
+                                                           String pageToken, Integer pageSize,
+                                                           SortOrder orderBy, String nameFilter
+      ) {
+
+    if (pageToken != null && pageToken.equals("0")) {
+      return new PaginatedApplicationRecords(Collections.emptyList(), "0");
+    }
+
+    Comparator<ApplicationRecord> compareByApplicationId = (orderBy == SortOrder.DESC)
+        ? Comparator.comparing(ApplicationRecord::getName).reversed()
+        : Comparator.comparing(ApplicationRecord::getName);
+
+    Integer startIndex = (Strings.isNullOrEmpty(pageToken)) ? 0 : Integer.parseInt(pageToken);
+    boolean isLastPage = (pageSize + startIndex >= apps.size());
```

Comment:
This won't work with filter. You can do easier. Just prepare a page and then if it's less then required - it's last page. It's fine to give a next 0-size page if you have exactly needed number of elements

Reply:
No need of `endIndex` or `isLastPage` this way, just `.skip(startIndex).limit(pageSize)`